src/newwave/main.mo

```diff
@@ -10,1382 +10,598 @@ import Iter "mo:base/Iter";
+    switch (result) {
+      case (null) { return false };
+      case (entity) { return true };
     };
   };
 
-  func getEntityByAttribute(attribute : Text, attributeValue : Text) : ?Entity.Entity {
-    switch(attribute) {
-      case "internalId" {
-        return getEntity(attributeValue);
-      };
-      case "externalId" {
-        for ((k, entity) in entitiesStorage.entries()) {
-          if (entity.externalId == ?attributeValue) {
-            return ?entity;
-          };          
+  /**
+  * Function takes in an entity update object and updates the entity that corresponds with the update
+  * if it exists and the caller has permissions. Otherwise an error is returned
+  *
+  * @return Either the Entity ID if the call was successful or an error if not
+  */
+  private func updateEntity(caller : Principal, entityUpdateObject : Entity.EntityUpdateObject) : async Entity.EntityIdResult {
+    var entity = getEntity(entityUpdateObject.id);
+    switch (entity) {
+      case null { return #Err(#EntityNotFound) };
+      case (?entityToUpdate) {
+        switch (Principal.equal(entityToUpdate.owner, caller)) {
+          case false {
+            return #Err(#Unauthorized);
+          };
+          // Only owner may update the Entity
+          case true {
+            let updatedEntity : Entity.Entity = Entity.updateEntityFromUpdateObject(entityUpdateObject, entityToUpdate);
+            let result = putEntity(updatedEntity);
+            return #Ok(updatedEntity.id);
+          };
         };
-        return null;
       };
-      case _ { return null; }
     };
   };
 
-  func getEntitiesByAttribute(attribute : Text, attributeValue : Text) : [Entity.Entity] {
-    switch(attribute) {
-      case "externalId" {
-        var entitiesToReturn = List.nil<Entity.Entity>();
-        for ((k, entity) in entitiesStorage.entries()) {
-          if (entity.externalId == ?attributeValue) {
-            entitiesToReturn := List.push<Entity.Entity>(entity, entitiesToReturn);
-          };          
-        };
-        return List.toArray<Entity.Entity>(entitiesToReturn);
-      };
-      case _ { return []; }
-    };
-  };
+  /**
+   * Function takes in a caller and the bridge id and attempts to delete the bridge. If the caller is the bridge owner,
+   * the bridge will be deleted as long as the reference within the entity will be deleted
+   *
+   * @return The Bridge id of the deleted bridge or an error
+  */
+  func deleteEntity(caller : Principal, entityId : Text) : async Entity.EntityIdResult {
+    switch (getEntity(entityId)) {
+      case null { return #Err(#EntityNotFound) };
+      case (?entityToDelete) {
+        switch (Principal.equal(entityToDelete.owner, caller)) {
+          case false {
+            return #Err(#Unauthorized);
+          }; // Only owner may delete the Entity
+          case true {
+            // First delete all the bridges pointing to this Entity
+            for (toBridge in entityToDelete.toIds.vals()) {
```

Comment:
putting this link to code as a reference here, how we can parallelize this in the future which will become relevant when multiple canisters are involved: https://github.com/patnorris/NewWavePre/blob/a023d6605ce0ad5f4f2dd93b4bbeb65570997e27/NewWaveV0/0.0.2/src/newwave/entity_retriever.mo#L34

Reply:
good to know. Long term good data synchronization will be key